libraries/ESP8266WiFi/src/WiFiClientSecureBearSSL.cpp

```diff
@@ -461,8 +458,19 @@ int WiFiClientSecure::_run_until(unsigned target, bool blocking) {
       unsigned char *buf;
       size_t len;
       int wlen;
+      size_t availForWrite;
 
       buf = br_ssl_engine_sendrec_buf(_eng, &len);
+      availForWrite = WiFiClient::availableForWrite();
+      
+      if (!blocking && len > availForWrite) {
+        /* 
+           writes on WiFiClient will block if len > availableForWrite()
+           this is needed to prevent available() calls from blocking
+           on dropped connections 
+        */
+        len = availForWrite;
```

Comment:
In that case WiFiClient::write just returns 0 instantly:
```cpp
size_t WiFiClient::write(const uint8_t *buf, size_t size)
{
    if (!_client || !size)
    {
        return 0;
    }
    _client->setTimeout(_timeout);
    return _client->write(buf, size);
}
```

Previously if would return 0 after timeout.
_run_until then returns -1

There is (and was) a minor issue here: because _run_until first tries to send record data, and returns -1 if it can't be sent, we never get to test other states.  It's possible (I think) that there is actually app data to receive (i.e. state & BR_SSL_RECVAPP), or data in the WiFiClient input buffer to be processed, if it wasn't read out before write attempts fill the tx buffer on a dropped connection.

The return -1 here could probably be removed, and the "no_work = 0; continue;" statements moved up into the "if (wlen > 0) block", then the loop allowed to run through the other states before breaking at the bottom, but it would need a lot more testing.

As it stands the "if (wlen > 0)" on line 485 is actually redundant, because we only get there if wlen > 0

Reply:
When `availForWrite`==0, -1 is returned from `WiFiClientSecure::_run_until`.
Should it retry until timeout ?

Reply:
IMO, no, not when blocking==false. 

Turns out the complex ordering of comparisons in _run_until is only needed in half-duplex mode.  In full duplex it can be simplified, as in this [example:](https://www.bearssl.org/gitweb/?p=BearSSL;a=blob;f=tools/sslio.c;h=e784d07877b98f395edb4348cc6ec7dec02f8928;hb=f3ac21cb00646600475b02ecd2c592ee36e417da#l104)

I've been experimenting with getting rid of _run_until entirely due to a few issues I've had with it (this one, https://github.com/esp8266/Arduino/issues/6464, and the sometimes-double or triple timeout).  The underlying IO can be broken out into _readRecordData and _writeRecordData functions, which to me seems simpler and more natural.  My experimental version is here for the curious:

https://github.com/johnm545/Arduino/blob/rewrite-brssl-io/libraries/ESP8266WiFi/src/WiFiClientSecureBearSSL.cpp

Reply:
That last sounds like something for @earlephilhower

Reply:
Looks really cool, @johnm545!  Please do think about polishing up to another PR once you've got client and server tests running.  One less way of having a false/artificial deadlock would be great, and fixing a hard-to-repro bug.  Good stuff if it's kosher!

Client testing can be run on the host, so maybe it's worthwhile simulating the network failures/hiccups you're seeing on the real system there.  Sort of a fault injection framework...

Reply:
Thanks, I'm running this branch in my projects, but have yet to get the tests setup.  Testing on the host is a good idea - it should be possible to write some host tests that simulate connection drops etc.